From 4dd5ea93df2863d98c272851c740bceb2a61207f Mon Sep 17 00:00:00 2001
From: Jaroslav Kysela <jkysela@redhat.com>
Date: Wed, 3 Sep 2014 07:31:38 -0400
Subject: [alsa] usb-audio: Fix deadlocks at resuming

Message-id: <140972823573-286-patchset-jkysela@redhat.com>
Patchwork-id: 90982
O-Subject: [RHEL 7.1 PATCH 285/442] ALSA: usb-audio: Fix deadlocks at resuming
Bugzilla: 1112200
RH-Acked-by: John Linville <linville@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Ivan Vecera <ivecera@redhat.com>

The recent addition of the USB audio mixer suspend/resume may lead to
deadlocks when the driver tries to call usb_autopm_get_interface()
recursively, since the function tries to sync with the finish of the
other calls.  For avoiding it, introduce a flag indicating the resume
operation and avoids the recursive usb_autopm_get_interface() calls
during the resume.

Reported-and-tested-by: Bryan Quigley <gquigs@gmail.com>
Signed-off-by: Takashi Iwai <tiwai@suse.de>

(picked from kernel: 1ee23fe07ee83a38ecee927e701f762888ada942)
rhbz#1112200
Signed-off-by: Jaroslav Kysela <jkysela@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/sound/usb/card.c b/sound/usb/card.c
index e769d39..c3b5b7d 100644
--- a/sound/usb/card.c
+++ b/sound/usb/card.c
@@ -651,7 +651,7 @@ int snd_usb_autoresume(struct snd_usb_audio *chip)
  int err = -ENODEV;
 
  down_read(&chip->shutdown_rwsem);
- if (chip->probing)
+ if (chip->probing && chip->in_pm)
   err = 0;
  else if (!chip->shutdown)
   err = usb_autopm_get_interface(chip->pm_intf);
@@ -663,7 +663,7 @@ int snd_usb_autoresume(struct snd_usb_audio *chip)
 void snd_usb_autosuspend(struct snd_usb_audio *chip)
 {
  down_read(&chip->shutdown_rwsem);
- if (!chip->shutdown && !chip->probing)
+ if (!chip->shutdown && !chip->probing && !chip->in_pm)
   usb_autopm_put_interface(chip->pm_intf);
  up_read(&chip->shutdown_rwsem);
 }
@@ -712,6 +712,8 @@ static int __usb_audio_resume(struct usb_interface *intf, bool reset_resume)
   return 0;
  if (--chip->num_suspended_intf)
   return 0;
+
+ chip->in_pm = 1;
  /*
   * ALSA leaves material resumption to user space
   * we just notify and restart the mixers
@@ -727,6 +729,7 @@ static int __usb_audio_resume(struct usb_interface *intf, bool reset_resume)
  chip->autosuspended = 0;
 
 err_out:
+ chip->in_pm = 0;
  return err;
 }
 
diff --git a/sound/usb/usbaudio.h b/sound/usb/usbaudio.h
index 25c4c7e..91d0380 100644
--- a/sound/usb/usbaudio.h
+++ b/sound/usb/usbaudio.h
@@ -40,6 +40,7 @@ struct snd_usb_audio {
  struct rw_semaphore shutdown_rwsem;
  unsigned int shutdown:1;
  unsigned int probing:1;
+ unsigned int in_pm:1;
  unsigned int autosuspended:1; 
  unsigned int txfr_quirk:1; /* Subframe boundaries on transfers */
  
-- 
1.7.1